lib/rust/ensogl/core/src/display/object/instance.rs

```diff
@@ -5,354 +5,535 @@
+        self.parent_bind.is_some()
     }
 
-    /// Parent object getter.
-    pub fn parent(&self) -> Option<Instance<Host>> {
-        self.parent_bind.borrow().as_ref().and_then(|t| t.parent())
+    /// Get reference to the parent object if any.
+    pub fn parent(&self) -> Option<Instance> {
+        self.parent_bind.parent()
     }
 
-    /// Count of children objects.
-    pub fn children_count(&self) -> usize {
-        self.children.borrow().len()
+    /// The index of this display object in the parent's children list.
+    fn my_index(&self) -> Option<usize> {
+        self.parent_bind.child_index()
     }
 
-    /// Recompute the transformation matrix of this object and update all of its dirty children.
-    pub fn update(&self, host: &Host) {
-        let origin0 = Matrix4::identity();
-        self.update_with_origin(host, origin0, false, false, None)
+    fn has_visible_parent(&self) -> bool {
+        self.parent_bind.parent().map_or(false, |parent| parent.is_visible())
     }
 
-    /// The default visibility of a new [`Instance`] is false. You can use this function to override
-    /// it. It is mainly used for a special 'root' element if such exists.
-    pub fn force_set_visibility(&self, visibility: bool) {
-        self.visible.set(visibility);
-        // TODO[ao] this function should make the next update call on_show or on_hide
-        //     https://github.com/enso-org/ide/issues/1406
+    /// Number of children of this object.
+    pub fn children_count(&self) -> usize {
+        self.children.borrow().len()
     }
 
-    /// Removes child by a given index. Does nothing if the index was incorrect.
-    fn remove_child_by_index(&self, index: usize) {
-        self.children.borrow_mut().remove(index).for_each(|child| {
-            child.upgrade().for_each(|child| child.unsafe_unset_parent_without_update());
-            self.dirty.children.unset(&index);
-            self.dirty.removed_children.set(child);
-        });
+    /// Removes and returns the parent bind. Please note that the parent is not updated as long as
+    /// the parent bind is not dropped.
+    fn take_parent_bind(&self) -> Option<ParentBind> {
+        let parent_bind = self.parent_bind.take_bind();
+        if let Some(parent) = parent_bind.as_ref().and_then(|t| t.parent.upgrade()) {
+            let is_focused = self.event.focused_descendant.borrow().is_some();
+            if is_focused {
+                parent.propagate_up_no_focus_instance();
+            }
+        }
+        parent_bind
+    }
+
+    /// Set parent of the object. If the object already has a parent, the parent would be replaced.
+    fn set_parent_bind(&self, bind: ParentBind) {
+        trace!("Adding new parent bind.");
+        if let Some(parent) = bind.parent() {
+            if let Some(focus_instance) = &*self.event.focused_descendant.borrow() {
+                parent.blur_tree();
+                parent.propagate_up_new_focus_instance(focus_instance);
+            }
```

Comment:
I'd take the old parent bind first, so `propagate_up_no_focus_instance` will be updated before the `propagate_up_new_focus_instance`, so there will be not where two separate branches have "focused descendant" state.

Reply:
I like it a lot. Changed.